extension/include/boost/di/extension/providers/runtime_provider.hpp

```diff
@@ -0,0 +1,123 @@
+//
+// Copyright (c) 2012-2018 Kris Jusiak (kris at jusiak dot net)
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+#pragma once
+
+#include <cassert>
+#include <functional>
+#include <memory>
+#include <typeindex>
+#include <unordered_map>
+
+#include "boost/di.hpp"
+
+BOOST_DI_NAMESPACE_BEGIN
+namespace extension {
+
+class runtime_provider : public config {
+  using bindings_t = std::unordered_map<std::type_index, std::function<void *()>>;
+
+  class abstract_provider {
+   public:
+    explicit abstract_provider(const bindings_t &bindings) : bindings_{bindings} {}
+
+    template <class...>
+    struct is_creatable {
+      static constexpr auto value = true;
+    };
+
+    template <class T, class TInitialization, class TMemory, class... TArgs>
+    auto get(const TInitialization &, const TMemory &, TArgs &&... args) const {
+      auto it = bindings_.find(std::type_index(typeid(T)));
+      if (it == bindings_.end()) {
+        return get<T>(std::integral_constant < bool,
+                      !std::is_abstract<T>::value && std::is_constructible<T, TArgs...>::value > {},
+                      std::forward<TArgs>(args)...);
+      }
+      return static_cast<T *>(it->second());
+    }
+
+   private:
+    template <class T, class... TArgs>
+    auto get(std::true_type, TArgs &&... args) const {
+      return new T{std::forward<TArgs>(args)...};
+    }
+
+    template <class T, class... TArgs>
+    auto get(std::false_type, TArgs &&...) const {
+      assert(false && "Type not bound!");
+      return static_cast<T *>(nullptr);
+    }
+
+    const bindings_t &bindings_;
+  };
+
+ public:
+  auto provider(...) { return abstract_provider{bindings_}; }
+  auto &bindings() { return bindings_; }
+
+ private:
+  bindings_t bindings_{};
+};
+
+class injector : public core::injector<runtime_provider> {
```

Comment:
Would be great if we could create this injector from existing non-runtime one?

Reply:
That will be possible with install(injector). Example is coming in a new MR.